mathjax3-ts/output/chtml/Wrappers/mpadded.ts

```diff
@@ -0,0 +1,138 @@
+        const [H, D, W, dh, dd, dw, x, y] = this.getDimens();
+        //
+        // If the width changed, set the width explicitly
+        //
+        if (dw) {
+            style.width = this.em(W + dw);
+        }
+        //
+        // If the height or depth changed, use margin to make the change
+        //
+        if (dh || dd) {
+            style.margin = this.em(dh) + ' 0 ' + this.em(dd);
+        }
+        //
+        // If there is a horizotnal or vertical shift,
+        //   use relative poisitioning to move the contents
+        //
+        if (x || y) {
+            style.position = 'relative';
+            content.push(this.html('mjx-rbox', {style: {left: this.em(x), top: this.em(-y)}}));
+        }
+        //
+        //  Create the HTML with the proper styles and content
+        //
+        chtml = chtml.appendChild(this.html('mjx-block', {style: style}, content));
+        for (const child of this.childNodes) {
+            child.toCHTML(chtml.firstChild as HTMLElement || chtml);
+        }
+    }
+
+    /*
+     * Get the content bounding box, and the change in size and offsets
+     *   as specified by the parameters
+     *
+     * @return{number[]}  The original height, depth, width, the changes in height, depth,
+     *                    and width, and the horizontal and vertical offsets of the content
+     */
+    protected getDimens() {
+        const values = this.node.attributes.getList('width', 'height', 'depth', 'lspace', 'voffset');
+        const bbox = super.computeBBox();  // get unmodified bbox of children
+        let {w, h, d} = bbox;
+        let W = w, H = h, D = d, x = 0, y = 0;
+        if (values.width !== '')   w = this.dimen(values.width, bbox, 'w', 0);
+        if (values.height !== '')  h = this.dimen(values.height, bbox, 'h', 0);
+        if (values.depth !== '')   d = this.dimen(values.depth, bbox, 'd', 0);
+        if (values.voffset !== '') y = this.dimen(values.voffset, bbox);
+        if (values.lspace !== '')  x = this.dimen(values.lspace, bbox);
+        return [H, D, W, h - H, d - D, w - W, x, y];
+    }
+
+    /*
+     * Get a particular dimension, which can be relative to any of the BBox dimensions,
+     *   and can be an offset from the default size of the given dimension.
+     *
+     * @param{Property} length   The value to be converted to a length in ems
+     * @param{BBox} bbox         The bbox of the mpadded content
+     * @param{string} d          The default dimension to use for relative sizes ('w', 'h', or 'd')
+     * @param{number} m          The minimum value allowed for the dimension
+     * @return{number}           The final dimension in ems
+     */
+    protected dimen(length: Property, bbox: BBox, d: string = '', m: number = null) {
+        length = String(length);
+        const match = length.match(/width|height|depth/);
+        const size = (match ? (bbox as IBBox)[match[0].charAt(0)] : (d ? (bbox as IBBox)[d] : 0)) as number;
```

Comment:
If this is the only reason for IBBox, the type definition should be sufficient.
However, it's still very confusing. I believe we should have a enum for the properties h,d,w of a bbox and use this in the definition of  the BBoxData type.

Reply:
Taken care of via `bbox[match[0].charAt(0) as (keyof BBox)]`.